src/services/mqtt/mqtt.service.ts

```diff
@@ -0,0 +1,105 @@
+﻿import { Injectable } from "@nestjs/common";
+import { MQTTBrokerDevice } from "@entities/mqtt-broker-device.entity";
+import { random } from "crypto-js/lib-typedarrays";
+import { algo, enc, PBKDF2 } from "crypto-js";
+import { ApplicationService } from "@services/device-management/application.service";
+import { AuthenticationType } from "@enum/authentication-type.enum";
+import * as fs from "fs";
+import { createCertificate, createCSR, createPrivateKey } from "pem";
+import { caCertPath, caKeyPath } from "@resources/resource-paths";
+
+@Injectable()
+export class MqttService {
+    constructor(private applicationService: ApplicationService) {}
+
+    public async generateCertificate(deviceName: string): Promise<CertificateDetails> {
+        const certificateDetails = new CertificateDetails();
+        try {
+            createPrivateKey(2048, (err, { key }) => {
+                if (err) {
+                    console.log("keyerr", err);
+                    return;
+                }
+                certificateDetails.deviceCertificateKey = key;
+                createCSR(
+                    {
+                        commonName: deviceName,
+                        clientKey: key,
+                    },
+                    function (err, { csr }) {
+                        if (err) throw err;
+                        const caCert = fs.readFileSync(caCertPath);
+                        const caKey = fs.readFileSync(caKeyPath).toString();
+                        createCertificate(
+                            {
+                                csr: csr,
+                                serviceKey: caKey,
+                                serviceCertificate: caCert,
+                                serial: Date.now(),
+                                days: 365,
+                                serviceKeyPassword: process.env.CA_KEY_PASSWORD || "test",
+                            },
+                            async function (err, cert) {
+                                if (err) throw err;
+
+                                certificateDetails.deviceCertificate = cert.certificate;
+                                certificateDetails.ca = caCert.toString();
+                            }
+                        );
+                    }
+                );
+            });
+            while (certificateDetails.deviceCertificate === undefined) {
```

Comment:
Maybe insert a comment why this is necessary

Reply:
Done